pkg/virtual/framework/dynamic/apiserver/openapi.go

```diff
@@ -0,0 +1,383 @@
+	servicesLock     keymutex.KeyMutex
+	openAPIV3Service *handler3.OpenAPIService
+}
+
+func newOpenAPIHandler(
+	apiSetRetriever apidefinition.APIDefinitionSetGetter,
+	goRestfulContainer *restful.Container,
+	openapiv3Config *common.OpenAPIV3Config,
+	serviceCacheSize int,
+	delegate http.Handler) *openAPIHandler {
+	openAPIV3Service := handler3.NewOpenAPIService()
+
+	return &openAPIHandler{
+		apiSetRetriever:    apiSetRetriever,
+		goRestfulContainer: goRestfulContainer,
+		openapiv3Config:    openapiv3Config,
+		delegate:           delegate,
+
+		services:         lru.New(serviceCacheSize),
+		servicesLock:     keymutex.NewHashed(0),
+		openAPIV3Service: openAPIV3Service,
+	}
+}
+
+func (o *openAPIHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	log := klog.FromContext(ctx).WithValues("path", req.URL.Path)
+	apiDomainKey := dyncamiccontext.APIDomainKeyFrom(ctx)
+
+	// Get all available APIDefinitions for this Virtual Workspace
+	apiSet, hasLocationKey, err := o.apiSetRetriever.GetAPIDefinitionSet(ctx, apiDomainKey)
+	if err != nil {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewInternalError(fmt.Errorf("unable to determine API definition set: %w", err)),
+			errorCodecs, schema.GroupVersion{},
+			w, req)
+		return
+	}
+	if !hasLocationKey {
+		o.delegate.ServeHTTP(w, req)
+		return
+	}
+
+	// Get the OpenAPI service from cache or create a new one
+	key, err := apiConfigurationKey(apiSet)
+	if err != nil {
+		responsewriters.InternalError(w, req, err)
+		return
+	}
+	log = log.WithValues("key", key)
+
+	// We want to make sure we don't generate OpenAPI v3 spec if not necessary because it's an expensive operation.
+	// Putting the generated OpenAPI v3 service in a cache using the configuration key generated above solves this in
+	// the most of cases.
+	// However, it doesn't cover the case where we have two concurrent requests with the same caching key, in which
+	// we would generate the OpenAPI v3 spec two times for the same key (if the spec is not cached already).
+	// That's why we have a "double-locking mechanism":
+	//   1) we store a special struct in a cache that contains a channel and an OpenAPI v3 service; if the channel is
+	//      closed, we know that we generated the OpenAPI v3 spec for that key and we can just serve it
+	// 	 2) that struct is supposed to be a singleton for a given caching key, and because of that, we need some
+	//      guarantee that no two threads are going to try to create an instance of that struct for the same key
+	//      (that's why we have servicesLock)
+	o.servicesLock.LockKey(key)
+	entry, ok := o.services.Get(key)
+	var item *openAPIServiceItem
+
+	// If we found the struct, and the channel is closed, and the spec was generated successfully, serve the already
+	// generated spec, otherwise regenerate it.
+	if ok {
+		item = entry.(*openAPIServiceItem)
+		if item != nil {
+			_, ok := <-item.done
+			if !ok && item.err == nil {
```

Comment:
why the ok check? Why not just `<- item.done`?

Reply:
I thought for some reason that `<-item.done` is blocking even if the channel is closed, but now that I tested it again, turns out it isn't blocking. I'll change that.